File: app/src/main/java/me/zhanghai/android/files/navigation/NavigationItem.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.navigation;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.storage.StorageVolume;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.content.res.AppCompatResources;
import java8.nio.file.Path;

public abstract class NavigationItem {

    public abstract long getId();

    @NonNull
    public Drawable getIcon(@NonNull Context context) {
        return AppCompatResources.getDrawable(context, getIconRes());
    }

    @DrawableRes
    protected abstract int getIconRes();

    @NonNull
    public String getTitle(@NonNull Context context) {
        return context.getString(getTitleRes());
    }

    @StringRes
    protected abstract int getTitleRes();

    @Nullable
    public String getSubtitle(@NonNull Context context) {
        return null;
    }

    public boolean isChecked(@NonNull Listener listener) {
        return false;
    }

    public abstract void onClick(@NonNull Listener listener);

    public boolean onLongClick(@NonNull Listener listener) {
        return false;
    }

    interface Listener {

        @NonNull
        Path getCurrentPath();

        void navigateTo(@NonNull Path path);

        void navigateToRoot(@NonNull Path path);

        void onAddDocumentTree();

        void onRemoveDocumentTree(@NonNull Uri treeUri, @Nullable StorageVolume storageVolume);

        void onEditBookmarkDirectory(@NonNull BookmarkDirectory bookmarkDirectory);

        void closeNavigationDrawer();

        void startActivity(@NonNull Intent intent);
    }
}

File: app/src/main/java/me/zhanghai/android/files/navigation/NavigationItem.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.navigation

import android.content.Context
import android.content.Intent
import android.graphics.drawable.Drawable
import androidx.annotation.DrawableRes
import java8.nio.file.Path
import me.zhanghai.android.files.compat.getDrawableCompat

abstract class NavigationItem {
    abstract val id: Long

    fun getIcon(context: Context): Drawable = context.getDrawableCompat(iconRes!!)

    @get:DrawableRes
    protected abstract val iconRes: Int?

    abstract fun getTitle(context: Context): String

    open fun getSubtitle(context: Context): String? = null

    open fun isChecked(listener: Listener): Boolean = false

    abstract fun onClick(listener: Listener)

    open fun onLongClick(listener: Listener): Boolean = false

    interface Listener {
        val currentPath: Path
        fun navigateTo(path: Path)
        fun navigateToRoot(path: Path)
        fun launchIntent(intent: Intent)
        fun closeNavigationDrawer()
    }
}
